insights/terraform/projects/bfd/README.md

```diff
@@ -16,51 +16,58 @@ table. Most other parts of this project will depend upon API-Requests.
+be clear. For example, we have `module.glue-table-api-requests` and
+`aws_glue_crawler.glue-crawler-api-requests`.
+
+## Adding new columns
+
+The `api_requests` table has hard-coded column fields, which is unavoidable due to limitations in
+Kinesis Firehose's format_conversion feature.
+
+Based on historical log files, this list is meant to contain every field ever used *so far*.
+However, when new fields are added to the FHIR server's log files, they will also need to be added
+to api-requests/glue.tf and then the tables will need to be updated in AWS. The procedure is:
+
+1. Add the new column to terraform.
+2. Run terraform apply.
+3. To verify that the new column has been added, wait about fifteen minutes to be sure that some
+log files have been processed via Kinesis Firehose, and then run a quick Athena query such as:
+
+```sql
+SELECT
+  "cw_id",
+  "cw_timestamp",
+  "<new_field>"
+FROM
+  "bfd_insights_bfd_<underscore_environment>_api_requests"
+WHERE
+  "<new_field>" IS NOT NULL
+LIMIT 50;
+```
 
-```mermaid
-flowchart TD
-    CloudWatch["CloudWatch: Historical Logs"] -->|Manual Export| S3
-    S3["S3 Bucket"] -->|Crawler: History| History["Glue Table: API History"]
-    History -->|Glue Job: History Ingest| APIRequests["Glue Table: API Requests"]
+## Manual Ingestion of Log Files
 
-    EC2["CloudWatch Log Subscription (Real-Time)"] --> Firehose["Kinesis Firehose"]
-    Firehose -->|Lambda| History
-```
+This process will be done via a series of complex Athena queries, in order to reduce the amount of
+AWS Glue we have to perform. This approach is far more cost-effective and faster.
 
-### Manual Ingestion of Log Files
-
-Note: Replace `<environment>` and `<account-number>` with the name of your environment, such as
-`prod` or `prod-sbx`, and the AWS account number, respectively. Replace any `-` with `_` in
-`<underscore_environment>` (Athena doesn't like hyphens in table names).
-
-1. CloudWatch > Log Groups > `/bfd/<environment>/bfd-server/access.json`
-    - Actions > Export Data to Amazon S3
-        - Choose time period
-        - Select Account: *This Account*
-        - S3 Bucket Name: `bfd-insights-bfd-app-logs`
-        - S3 Bucket Prefix: `history/temp-<environment>`.
-    - Export. This took about 80 minutes for 3 weeks of prod-sbx logs.
-
-2. S3
-    - Select bucket `bfd-insights-bfd-app-logs`
-    - Go to path `history/temp-<environment>`
-    - Select all objects *except* `aws-logs-write-test`
-    - Actions > Move
-        - Bucket
-        - Destination: `bfd-insights-bfd-<account-number>/databases/bfd-insights-bfd-<environment>/bfd_insights_bfd_<underscore_environment>_api_history/`
-        - Move
-
-3. AWS Glue > Workflows > `bfd-insights-bfd-<environment>-history-workflow`
-    - Actions > Run. The entire workflow may take a bit to run through, but you can track progress in the graph: History > (choose the top item) > Run Details.
+**TODO: Add a reference to the runbook once completed.**
```

Comment:
So is the creation of a runbook a separate ticket?

Reply:
Yep: it's [2131](https://jira.cms.gov/browse/BFD-2131), assigned to @christrotteradhoc